src/lib.rs

```diff
@@ -1,14 +1,180 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+mod args;
+mod context;
+mod rule;
+mod violation;
+
+use std::{
+    borrow::Cow,
+    process,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+pub use args::Args;
+use clap::Parser;
+use context::QueryMatchContext;
+use rule::{ResolvedRule, Rule, RuleBuilder, RuleListenerBuilder};
+use tree_sitter::Query;
+use violation::ViolationBuilder;
+
+use crate::context::Context;
+
+#[macro_export]
+macro_rules! regex {
+    ($re:expr $(,)?) => {{
+        static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
+        RE.get_or_init(|| regex::Regex::new($re).unwrap())
+    }};
+}
+
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP: &str = "_tree_sitter_lint_capture";
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT: &str = "@_tree_sitter_lint_capture";
+
+pub fn run(_args: Args) {
+    let language = tree_sitter_rust::language();
+    let context = Context::new(language);
+    let resolved_rules = get_rules()
+        .into_iter()
+        .map(|rule| rule.resolve(&context))
+        .collect::<Vec<_>>();
+    let aggregated_queries = AggregatedQueries::new(&resolved_rules, &context);
+    let tree_sitter_grep_args = tree_sitter_grep::Args::parse_from([
+        "tree_sitter_grep",
+        "-q",
+        &aggregated_queries.query_text,
+        "-l",
+        "rust",
+        "--capture",
+        CAPTURE_NAME_FOR_TREE_SITTER_GREP,
+    ]);
+    let reported_any_violations = AtomicBool::new(false);
+    tree_sitter_grep::run_with_callback(
+        tree_sitter_grep_args,
+        |capture_info, file_contents, path| {
+            let (rule_index, rule_listener_index) =
+                aggregated_queries.pattern_index_lookup[capture_info.pattern_index];
```

Comment:
Here we de-multiplex to resolve the `tree-sitter-grep`-reported match to one of our rule listeners

This general approach (of just running a single aggregated tree-sitter query (of all of the individual listener queries) once and letting `tree-sitter-grep` do its normal "single-capture-oriented" matching) relies on the underlying tree-sitter API reporting the same tree-sitter AST node multiple times if it matched multiple of the top-level patterns (because multiple listeners might be "interested" in the same tree-sitter AST node)

And it does appear to behave that way (I did a little separate sanity-check of that)

At that point this seems pretty efficient, we just crunch through the tree-sitter AST once per file